samza-sql-shell/src/main/java/org/apache/samza/sql/client/impl/SamzaExecutor.java

```diff
@@ -0,0 +1,582 @@
+        SamzaSqlApplicationRunner runner = m_executions.get(execId);
+        if (runner == null) {
+            return null;
+        }
+        return queryExecutionStatus(runner);
+    }
+
+    @Override
+    public String getErrorMsg() {
+        return m_lastErrorMsg;
+    }
+
+    @Override
+    public List<SqlFunction> listFunctions(ExecutionContext context) {
+        /**
+         * TODO: currently the Shell only shows some UDFs supported by Samza internally. We may need to require UDFs
+         *       to provide a function of getting their "SamzaSqlUdfDisplayInfo", then we can get the UDF information from
+         *       SamzaSqlApplicationConfig.udfResolver(or SamzaSqlApplicationConfig.udfMetadata) instead of registering
+         *       UDFs one by one as below.
+         */
+        List<SqlFunction> udfs = new ArrayList<>();
+        udfs.add(new SamzaSqlUdfDisplayInfo("RegexMatch", "Matches the string to the regex",
+            Arrays.asList(SamzaSqlFieldType.createPrimitiveFieldType(SamzaSqlFieldType.TypeName.STRING),
+                SamzaSqlFieldType.createPrimitiveFieldType(SamzaSqlFieldType.TypeName.STRING)),
+                SamzaSqlFieldType.createPrimitiveFieldType(SamzaSqlFieldType.TypeName.BOOLEAN)));
+
+        return udfs;
+    }
+
+    static void saveOutputMessage(OutgoingMessageEnvelope messageEnvelope) {
+        m_outputData.add(messageEnvelope);
+    }
+
+    private String getColumnTypeName(SamzaSqlFieldType fieldType) {
+        if (fieldType.isPrimitiveField()) {
+            return fieldType.getTypeName().toString();
+        } else if (fieldType.getTypeName() == SamzaSqlFieldType.TypeName.MAP) {
+            return String.format("MAP(%s)", getColumnTypeName(fieldType.getValueType()));
+        } else if (fieldType.getTypeName() == SamzaSqlFieldType.TypeName.ARRAY) {
+            return String.format("ARRAY(%s)", getColumnTypeName(fieldType.getElementType()));
+        } else {
+            SqlSchema schema = fieldType.getRowSchema();
+            List<String> fieldTypes = IntStream.range(0, schema.getFieldCount())
+                    .mapToObj(i -> schema.getFieldName(i) + " " + schema.getFieldTypeName(i))
+                    .collect(Collectors.toList());
+            String rowSchemaValue = Joiner.on(", ").join(fieldTypes);
+            return String.format("STRUCT(%s)", rowSchemaValue);
+        }
+    }
+
+    private SqlSchema convertAvroToSamzaSqlSchema(String schema) {
+        Schema avroSchema = Schema.parse(schema);
+        return getSchema(avroSchema.getFields());
+    }
+
+    private SqlSchema getSchema(List<Schema.Field> fields) {
+        SqlSchemaBuilder schemaBuilder = SqlSchemaBuilder.builder();
+        for (Schema.Field field : fields) {
+            schemaBuilder.addField(field.name(), getColumnTypeName(getFieldType(field.schema())));
+        }
+        return schemaBuilder.toSchema();
+    }
+
+    private SamzaSqlFieldType getFieldType(org.apache.avro.Schema schema) {
```

Comment:
Looks like this is avro to Samza SQL Schema converter. Can we refactor this and all other methods related to this conversion to a separate file?

Reply:
Yes. Done.